lib/compile/jtd/serialize.ts

```diff
@@ -228,8 +228,17 @@ function serializeString({gen, data}: SerializeCxt): void {
   gen.add(N.json, _`${useFunc(gen, quote)}(${data})`)
 }
 
-function serializeNumber({gen, data}: SerializeCxt): void {
-  gen.add(N.json, _`"" + ${data}`)
+function serializeNumber({gen, data, self}: SerializeCxt): void {
+  const condition = _`${data} === Infinity || ${data} === -Infinity || Number.isNaN(${data})`
+  const addNumber = (): CodeGen => gen.add(N.json, _`"" + ${data}`)
+
+  if (self.opts.specialNumbers === "null") {
+    gen.if(condition, () => gen.add(N.json, _`null`), addNumber)
+  } else if (self.opts.specialNumbers === "string") {
+    gen.if(condition, () => gen.add(N.json, str`"${data}"`), addNumber)
+  } else {
+    addNumber()
+  }
```

Comment:
```suggestion
  const sn = self.opts.specialNumbers
  if (sn === undefined || sn === "fast") {
    gen.add(N.json, _`"" + ${data}`)
  } else {
    gen.if(
      _`${data} === Infinity || ${data} === -Infinity || ${data} !== ${data}`,
      () => gen.add(N.json, sn === "null" ? _`null` : str`"${data}"`),
      () => gen.add(N.json, _`"" + ${data}`)
    )
  }
```

Reply:
We discussed in person, we will still have the 3 options but I got rid of the wrapping function `addNumber` and added in the default handling.